module/zfs/mmp.c

```diff
@@ -125,27 +130,79 @@ mmp_thread_stop(dsl_pool_t *dp)
 	mmp->mmp_thread_exiting = 0;
 }
 
+vdev_t *
+vdev_random_leaf(spa_t *spa)
+{
+	vdev_t *vd, *child;
+
+	ASSERT(spa);
+	ASSERT(spa_config_held(spa, SCL_STATE, RW_READER) == SCL_STATE);
+
+	vd = spa->spa_root_vdev;
+	while (!vd->vdev_ops->vdev_op_leaf) {
+		child = vd->vdev_child[spa_get_random(vd->vdev_children)];
+		if (!vdev_is_dead(child))
+			vd = child;
+	}
+
+	return (vd);
+}
+
 static void
 mmp_thread(dsl_pool_t *dp)
 {
 	mmp_thread_state_t *mmp = &dp->dp_mmp;
-	clock_t start;
-	/*
 	spa_t *spa = dp->dp_spa;
-	*/
+	abd_t *abd;
+	char *buf;
+	int bufsize = (1ULL << MAX_UBERBLOCK_SHIFT);
+
+	abd = abd_alloc(bufsize, B_TRUE);
+	buf = kmem_alloc(bufsize, KM_SLEEP);
+	bzero(buf, bufsize);
 
 	mmp_thread_enter(mmp);
 
 	dprintf("entered: pool %p\n", dp);
 
 	mmp->mmp_last_update = gethrtime();
 	for (;;) {
+		int flags = ZIO_FLAG_CONFIG_WRITER | ZIO_FLAG_CANFAIL;
+		clock_t start;
+		int l, n, rc;
+		vdev_t *vd;
+		zio_t *zio;
+
 		if (mmp->mmp_thread_exiting)
-			mmp_thread_exit(mmp, &mmp->mmp_thread);
+			goto cleanup_and_exit;
 
 		start = gethrtime();
 
-		/* Do actual mmp reads/writes here */
+		/* Copy the latest uberblock and update the time */
+		(void) memcpy(buf, &spa->spa_ubsync, sizeof (uberblock_t));
+		((uberblock_t *)buf)->ub_timestamp = gethrestime_sec();
+		abd_copy_from_buf(abd, buf, bufsize);
+
+		/*
+		 * Choose a random vdev and MMP block, and write over it in all
+		 * labels.  XXX need to retry if the zio fails
+		 */
+		spa_config_enter(spa, SCL_STATE, FTAG, RW_READER);
+		vd = vdev_random_leaf(spa);
+		if (vd) {
+			n = spa_get_random(MMP_BLOCKS_PER_LABEL);
+			zio = zio_root(spa, NULL, NULL, flags);
+			for (l = 0; l < VDEV_LABELS; l++)
+				vdev_mmp_write(zio, vd, l, n, abd, NULL, NULL,
```

Comment:
If the vdev_disk_physio() code is likely to work properly as is then that would be OK with me.  I don't want to introduce a new mechanism to submit I/O to disk and then have bugs in that new code doom MMP.

However it's not clear to me the zio pipeline creates a problem.  It seems to me like the classes of potential problems are either handled already or would be no better if we didn't use the pipeline:

1) Issues with the device, connection, adapters, device driver, etc.  These affect MMP whether it uses the ZFS pipeline or not.

2) Interference by the txg syncing process.   As long as I keep the MMP thread independent, ie no locking interaction with the sync thread, the concern is competition with the sync thread if it is pushing many writes through the pipeline.  MMP uses vdev_label_write() which uses the ZIO_PRIORITY_SYNC_WRITE flag.  This puts these writes ahead of all writes by the sync thread except for Uberblock writes, which themselves accomplish our goal.

3) Competition with sync writes and with ARC reads.  I believe using the flag ZIO_FLAG_DONT_QUEUE (credit Ahrens) with MMP writes will push them to the device ahead of those queued sync thread writes, and queued ARC reads.  I don't yet understand the related code, so I may be misunderstanding.

4) Bugs in the zio pipeline itself.  But a new mechanism is more likely to contain such bugs.

Am I mistaken?